ocamltest/main.ml

```diff
@@ -181,7 +181,55 @@ let test_file test_filename =
   (* Restore current working directory  *)
   Sys.chdir cwd
 
+let is_test s =
+  match tsl_block_of_file s with
+  | _ -> true
+  | exception _ -> false
+
+let ignored s =
+  s = "" || s.[0] = '_' || s.[0] = '.'
+
+let find_test_dirs dir =
+  let res = ref [] in
+  let rec loop dir =
+    let contains_tests = ref false in
+    Array.iter (fun s ->
+        if ignored s then ()
+        else begin
+          let s = Filename.concat dir s in
+          if Sys.is_directory s then loop s
+          else if not !contains_tests && is_test s then contains_tests := true
+        end
+      ) (Sys.readdir dir);
+    if !contains_tests then res := dir :: !res
+  in
+  loop dir;
+  List.rev !res
+
+let list_tests dir =
+  let res = ref [] in
+  if Sys.is_directory dir then begin
+    Array.iter (fun s ->
+        if ignored s then ()
+        else begin
+          let s' = Filename.concat dir s in
+          if Sys.is_directory s' || not (is_test s') then ()
+          else res := s :: !res
+        end
+      ) (Sys.readdir dir)
+  end;
+  List.rev !res
+
 let main () =
+  if !Options.find_test_dirs <> "" then begin
```

Comment:
I would expect `ocamltest -find-test-dirs foo -find-test-dirs bar` to work as expected, that is `Options.find_test_dirs` to be a list, and similarly for `Options.list_tests`. In fact, it seems a bit silly to invoke, for example, `ocamltest -find-test-dirs foo -list-tests bar`, but I don't see a reason to not support that either.

I would propose the following logic:
- list all test directories
- list all tests
- test all files to test
- if there were no test directories to list, no test to list, and no test to test, print usage

(The last point can be achieved with a `work_done : bool ref` flag, or written explicitly.)

Reply:
I agree and amended along suggested lines.